google/cloud/universe_domain_test.cc

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/universe_domain.h"
+#include "google/cloud/credentials.h"
+#include "google/cloud/internal/filesystem.h"
+#include "google/cloud/internal/oauth2_authorized_user_credentials.h"
+#include "google/cloud/internal/oauth2_google_application_default_credentials_file.h"
+#include "google/cloud/testing_util/scoped_environment.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include "google/cloud/universe_domain_options.h"
+#include <gmock/gmock.h>
+#include <fstream>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::google::cloud::testing_util::ScopedEnvironment;
+using ::testing::Eq;
+using ::testing::NotNull;
+
+auto constexpr kAuthorizedUserCredContents = R"""({
+  "client_id": "test-invalid-test-invalid.apps.googleusercontent.com",
+  "client_secret": "invalid-invalid-invalid",
+  "refresh_token": "1/test-test-test",
+  "type": "authorized_user"
+})""";
+
+std::string TempFileName() {
+  static auto generator =
+      google::cloud::internal::DefaultPRNG(std::random_device{}());
+  return google::cloud::internal::PathAppend(
+      ::testing::TempDir(),
+      ::google::cloud::internal::Sample(
+          generator, 16, "abcdefghijlkmnopqrstuvwxyz0123456789") +
+          ".json");
+}
+
+/// @test AuthorizedUser credentials do not support universe_domain and should
+/// always return the google default universe (GDU).
+TEST(AddUniverseDomainOption, GoogleDefaultCredentialsAuthorizedUser) {
+  auto const filename = TempFileName();
+  std::ofstream(filename) << kAuthorizedUserCredContents;
+  auto const env =
+      ScopedEnvironment(oauth2_internal::GoogleAdcEnvVar(), filename.c_str());
+  auto result_options = AddUniverseDomainOption(ExperimentalTag{}, Options{});
+  (void)std::remove(filename.c_str());
+
+  ASSERT_STATUS_OK(result_options);
+  ASSERT_TRUE(result_options->has<UnifiedCredentialsOption>());
```

Comment:
nit: this can be an `EXPECT_TRUE`... or we could just get rid of it, because `EXPECT_THAT(creds.get(), NotNull())` covers the same logic.